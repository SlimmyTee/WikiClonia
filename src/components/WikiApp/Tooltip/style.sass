.tooltip
  display: inline-block
  justify-content: center
  position: relative
  font-style: inherit


.tooltip-content
  position: absolute
  width: 40vw
  background-color: #121212
  text-align: left
  color: #fff
  transform: translateX(-50%)
  white-space: wrap
  border-radius: 4px
  padding: 6px 10px

  &:before
    content: ""
    position: absolute

.tooltip-content.top
  bottom: 100%
  left: 50%
  margin-bottom: 10px
  &:before
    top: 100%
    left: 50%
    margin-left: -8px
    border-left: 8px solid transparent
    border-right: 8px solid transparent
    border-top: 8px solid #000

