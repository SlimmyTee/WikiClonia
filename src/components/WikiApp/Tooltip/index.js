import React, { useState } from "react";
import { buildURL, summaryParams } from "./../../../WikiWrapper";
import Skeleton from "react-loading-skeleton";
import "./style.sass";

const extractSummary = (json, maxchar = 300) => {
  let content = json.query?.pages,
    extract = null,
    title = json.query?.normalized?.[0]?.to;
  for (const key in content) extract = content[key]?.extract;
  extract = extract.substr(0, maxchar) + "...";
  return { extract, title };
};

const Tooltip = props => {
  const { url, children } = props;
  const [show, setShow] = useState(false);
  const [summary, setSummary] = useState(null);
  const [title, setTitle] = useState(null);
  const handleMouseIn = () => setShow(true);
  const handleMouseOut = () => setShow(false);

  React.useEffect(() => {
    if (show && !summary) {
      fetch(buildURL(summaryParams(url)))
        .then(response => response.json())
        .then(json => {
          let { extract, title } = extractSummary(json);
          setSummary(extract);
          setTitle(title || url);
        })
        .catch(error => console.error(error));
    }
  }, [show]);

  return (
    <div
      className="tooltip"
      onMouseOver={handleMouseIn}
      onMouseLeave={handleMouseOut}
    >
      {show && (
        <div className="tooltip-content top">
          {title ? (
            <h3
              style={{ marginTop: 0, marginBottom: "5rem" }}
              className="tooltip-title"
            >
              {title}
            </h3>
          ) : (
            <Skeleton height={100} duration={1} />
          )}
          {summary}
        </div>
      )}

      {React.Children.toArray(children)}
    </div>
  );
};

export default Tooltip;
