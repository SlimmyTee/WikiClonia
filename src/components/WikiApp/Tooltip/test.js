import React from "react";
import Tooltip from "./";
import renderer from "react-test-renderer";

const titles = [
  "Pet_door",
  "New_York_City",
  "The_Last_Supper_(Leonardo)",
  "Leonardo_da_Vinci",
  "Mona_Lisa",
  "Renaissance"
];

for (const title of titles) {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <Tooltip url={title}>
          <a href="#">Test</a>
        </Tooltip>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
}
