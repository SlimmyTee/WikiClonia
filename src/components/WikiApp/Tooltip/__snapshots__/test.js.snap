// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="tooltip"
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
>
  <a
    href="#"
  >
    Test
  </a>
</div>
`;

exports[`renders correctly 2`] = `
<div
  className="tooltip"
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
>
  <a
    href="#"
  >
    Test
  </a>
</div>
`;

exports[`renders correctly 3`] = `
<div
  className="tooltip"
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
>
  <a
    href="#"
  >
    Test
  </a>
</div>
`;

exports[`renders correctly 4`] = `
<div
  className="tooltip"
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
>
  <a
    href="#"
  >
    Test
  </a>
</div>
`;

exports[`renders correctly 5`] = `
<div
  className="tooltip"
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
>
  <a
    href="#"
  >
    Test
  </a>
</div>
`;

exports[`renders correctly 6`] = `
<div
  className="tooltip"
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
>
  <a
    href="#"
  >
    Test
  </a>
</div>
`;
